src/sources/generic.rs

```diff
@@ -96,25 +98,44 @@ impl<F: AsRawFd> EventSource for Generic<F> {
     fn process_events<C>(
         &mut self,
         readiness: Readiness,
-        _: Token,
+        token: Token,
         mut callback: C,
     ) -> std::io::Result<()>
     where
         C: FnMut(Self::Event, &mut Self::Metadata) -> Self::Ret,
     {
+        if token != self.token {
+            return Ok(());
+        }
         callback(readiness, &mut self.file)
     }
 
-    fn register(&mut self, poll: &mut Poll, token: Token) -> std::io::Result<()> {
-        poll.register(self.file.as_raw_fd(), self.interest, self.mode, token)
+    fn register(
+        &mut self,
+        poll: &mut Poll,
+        token_factory: &mut TokenFactory,
+    ) -> std::io::Result<()> {
+        let token = token_factory.token();
+        poll.register(self.file.as_raw_fd(), self.interest, self.mode, token)?;
```

Comment:
I haven't fully thought this through, but maybe `[re]register()` should return the token on success?  Then you could do `self.token = poll.register(...)` which might be a bit more ergonomic. Of course, that's entirely a matter of opinion, some may prefer it to be more explicit. (Plus you'd still need to return `self.token` then.) Think about it.

Reply:
Well, given this method is invoked exactly once in `calloop`'s codebase (all other sources build upon `Generic`), and I expect it to be rarely directly invoked outside of it as well (again, most would just compose using the `Generic` source), I'm thinking I'd prefer to keep it like that, without trying to be smart with the design.

Reply:
Oh I see what you mean — even if I'm composing a source, I don't actually need my *own* token for the *composed* type, do I? This approach means I just recursively pass the token factory in to my sub-sources. Correct?

Reply:
Yes, exactly. And similarly, you just invoke `process_events` recursively unconditionnaly, and it is the inner source that compares with its tokens and decides if it actually has something to do.